underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java

```diff
@@ -0,0 +1,486 @@
+
+  // Not supported
+  @Override
+  public void setPermission(String path, String posixPerm) throws IOException {}
+
+  /**
+   * Treating S3 as a file system, checks if the parent directory exists.
+   * @param key the key to check
+   * @return true if the parent exists or if the key is root, false otherwise
+   */
+  private boolean parentExists(String key) {
+    // Assume root always has a parent
+    if (isRoot(key)) {
+      return true;
+    }
+    String parentKey = getParentKey(key);
+    return isFolder(parentKey);
+  }
+
+  /**
+   * Appends the directory suffix to the key.
+   * @param key the key to convert
+   * @return key as a directory path
+   */
+  private String convertToFolderName(String key) {
+    return key + FOLDER_SUFFIX;
+  }
+
+  /**
+   * Copies an object to another key.
+   * @param src the source key to copy.
+   * @param dst the destination key to copy to.
+   * @return true if the operation was successful, false otherwise
+   */
+  private boolean copy(String src, String dst) {
+    try {
+      src = stripPrefixIfPresent(src);
+      dst = stripPrefixIfPresent(dst);
+      LOG.info("Copying " + src + " to " + dst);
+      S3Object obj = new S3Object(dst);
+      mClient.copyObject(mBucketName, src, mBucketName, obj, false);
+      return true;
+    } catch (ServiceException se) {
+      LOG.error("Failed to rename file " + src + " to " + dst);
+      return false;
+    }
+  }
+
+  /**
+   * Internal function to delete a key in S3
+   * @param key the key to delete
+   * @return true if successful, false if an exception is thrown
+   */
+  private boolean deleteInternal(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        mClient.deleteObject(mBucketName, keyAsFolder);
+      } else {
+        mClient.deleteObject(mBucketName, stripPrefixIfPresent(key));
+      }
+    } catch (ServiceException se) {
+      LOG.error("Failed to delete " + key, se);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Gets the StorageObject representing the metadata of a key. If the key does not exist as a
+   * file or folder, null is returned
+   * @param key the key to get the object details of
+   * @return StorageObject of the key, or null if the key does not exist as a file or folder.
+   */
+  private StorageObject getObjectDetails(String key) {
+    try {
+      if (isFolder(key)) {
+        String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+        return mClient.getObjectDetails(mBucketName, keyAsFolder);
+      } else {
+        return mClient.getObjectDetails(mBucketName, stripPrefixIfPresent(key));
+      }
+    } catch (ServiceException se) {
+      return null;
+    }
+  }
+
+  /**
+   * Gets the parent key of the input key, or null if no parent exists
+   * @param key the key to get the parent of
+   * @return the the parent key
+   */
+  private String getParentKey(String key) {
```

Comment:
helper functions like this one and the next should be easily unit-tested, without requiring a S3 mock. Maybe add some unittests (in addition to integration tests) in next PR after unit-tests and integration tests are separated (TACHYON-432)